internal/storage/device_session.go

```diff
@@ -486,12 +486,12 @@ func GetDeviceSessionForPHYPayload(p *redis.Pool, phy lorawan.PHYPayload, txDR,
 		}
 
 		// the FCnt is valid, validate the MIC
-		macPL.FHDR.FCnt = fullFCnt
 		micOK, err := phy.ValidateUplinkDataMIC(s.GetMACVersion(), s.ConfFCnt, uint8(txDR), uint8(txCh), s.FNwkSIntKey, s.SNwkSIntKey)
 		if err != nil {
 			return DeviceSession{}, errors.Wrap(err, "validate mic error")
 		}
 		if micOK {
+			macPL.FHDR.FCnt = fullFCnt
```

Comment:
after pass validateUplinkDataMIC check,macPL.FHDR.FCnt = fullFCnt, can update the device's session fcnt in after task.